[now-ruby] Use require_relative for ruby sinatra test

m5o opened the pull request:
* to be equal with docs example
  * 🔗 https://github.com/zeit/docs/pull/1587

m5o committed:
use require_relative for ruby sinatra test

* to be equal with docs example
  * 🔗 https://github.com/zeit/docs/pull/1587

styfle reviewed (commented):
But why does it work with the original `require './app'`?

Or rather, why did the docs example not work?

m5o commented:
Are the integration tests in A different setup structure and served on a root directory instead of the `/api/` path? Sorry, no other idea. 
May the example from the docs worked one in a while 🤷‍♂ / previous ruby version etc. Just couldn‘t get it to work with just copy and paste. Do you need a example repo to proof?

styfle commented:
@m5o The docs example is using the `functions` property and the test here is using the (deprecated) `builds` property.

Perhaps the root directory vs api makes a difference like you said but my Ruby is not so good.

m5o commented:
Do you want me to update this example to the new `functions` property?

styfle commented:
> Do you want me to update this example to the new functions property?

Yes that would be great 👍 

I did some more digging and found that `require_relative` loads a file that is relative to the file containing the statement. However `require` with `./` indicates a path that is relative to your current working directory.

So it sounds like `require_relative` is a better option in most cases.

m5o committed:
use functions property instead of deprecated builds

* use functions property instead of deprecated builds
  * https://zeit.co/docs/configuration#project/functions

m5o committed:
create an api directory at the root of your project directory

> For all officially supported runtimes, the only requirement is to create an api directory at the root of your project directory, placing your Serverless Functions inside.

> WARNING: It is currently not possible to use the functions property when not using the /api directory for your Serverless Functions. If this behavior is required, you should use the builds property instead.

see https://zeit.co/docs/configuration#project/functions

m5o commented:
`nowJson.builds is not iterable` 🤔

It seems these integration tests require the `builds` configuration. 
Sorry I do not know anything about your build pipeline, can someone help?

styfle commented:
The tests for `@now/ruby` are not end-to-end. Meaning, the tests invoke the deployment API directly with `testDeployment.js` instead of deploying with Now CLI.

I think the fix is to change line 56 to fallback to an empty array if `builds` is undefined.

https://github.com/zeit/now/blob/aec8a7ac9a51b14f1f388f88ee98d573803e1e74/test/lib/deployment/test-deployment.js#L56

Something like:

```
  for (const build of nowJson.builds || []) {
```

styfle commented on packages/now-ruby/test/fixtures/05-sinatra/now.json:
```diff
@@ -1,5 +1,9 @@
+      "runtime": "@now/ruby@1.0.1"
+    }
+  },
   "probes": [{ "path": "/", "mustContain": "gem:RANDOMNESS_PLACEHOLDER" }]
```
This probe is wrong since you moved the file.

```suggestion
  "probes": [{ "path": "/api", "mustContain": "gem:RANDOMNESS_PLACEHOLDER" }]
```

styfle commented on packages/now-ruby/test/fixtures/05-sinatra/api/index.ru:
```diff
@@ -1,3 +1,3 @@
-require './app'
+require_relative 'app'
```
I think the `app.rb` file should not live in `/api` or else it will be turned into a serverless function.

Perhaps it should be `/src/app.rb` or maybe `/api/_lib/app.rb`?

m5o committed:
use /api path for probes

m5o committed:
rename app.rb

styfle commented:
Looks like you found a bug!

I'm seeing a `conflicting_file_path` error because it thinks that `Gemfile.lock` will be renamed to `Gemfile` even though these are not serverless function files.

I think the workaround is to move those two files back to the root instead of `/api`.

m5o committed:
move Gemfile* into root directory

* move Gemfile* into root directory

m5o committed:
revert functions property

it seems there are more issues with the functions property

* revert functions property
* keep changes minimal

kodiakhq[bot] committed:
Merge branch 'master' into ruby-sinatra-test

kodiakhq[bot] committed:
Merge branch 'master' into ruby-sinatra-test

m5o commented:
@coetry might not relevant anymore, because of the cleanup with https://github.com/zeit/docs/pull/1699 
What do you think?

styfle commented:
I'm going to close this, thanks!
